contracts/FungibleToken-v2.cdc

```diff
@@ -0,0 +1,197 @@
+/**
+
+# The Flow Fungible Token standard
+
+## `FungibleToken` contract
+
+The Fungible Token standard is no longer an interface
+that all fungible token contracts would have to conform to.
+
+If a users wants to deploy a new token contract, their contract
+does not need to implement the FungibleToken interface, but their tokens
+do need to implement the interfaces defined in this contract.
+
+## `Vault` resource interface
+
+Each fungible token resource type needs to implement the `Vault` resource interface.
+
+## `Provider`, `Receiver`, and `Balance` resource interfaces
+
+These interfaces declare pre-conditions and post-conditions that restrict
+the execution of the functions in the Vault.
+
+They are separate because it gives the user the ability to share
+a reference to their Vault that only exposes the fields functions
+in one or more of the interfaces.
+
+It also gives users the ability to make custom resources that implement
+these interfaces to do various things with the tokens.
+For example, a faucet can be implemented by conforming
+to the Provider interface.
+
+*/
+
+/// FungibleToken
+///
+pub contract FungibleToken {
+
+    /// Provider
+    ///
+    /// The interface that enforces the requirements for withdrawing
+    /// tokens from the implementing type.
+    ///
+    /// It does not enforce requirements on `balance` here,
+    /// because it leaves open the possibility of creating custom providers
+    /// that do not necessarily need their own balance.
+    ///
+    pub resource interface Provider {
+
+        /// withdraw subtracts tokens from the owner's Vault
+        /// and returns a Vault with the removed tokens.
+        ///
+        /// The function's access level is public, but this is not a problem
+        /// because only the owner storing the resource in their account
+        /// can initially call this function.
+        ///
+        /// The owner may grant other accounts access by creating a private
+        /// capability that allows specific other users to access
+        /// the provider resource through a reference.
+        ///
+        /// The owner may also grant all accounts access by creating a public
+        /// capability that allows all users to access the provider
+        /// resource through a reference.
+        ///
+        pub fun withdraw(amount: UFix64): @AnyResource{Vault} {
+            post {
+                // `result` refers to the return value
+                result.getBalance() == amount:
+                    "Withdrawal amount must be the same as the balance of the withdrawn Vault"
+            }
+        }
+    }
+    
+    pub resource interface Transferable {
+        /// Function for a direct transfer instead of having to do a deposit and withdrawal
+        ///
+        pub fun transfer(amount: UFix64, recipient: Address)
```

Comment:
I figured that people would only ever use this for a basic transfer, and if they want to use a special capability, they can just use the regular withdraw-deposit method. I'm not dedicated to that position though, so I could be convinced to put it as a capability if enough people want that.

Reply:
The problem with withdraw/deposit is that it is impossible to create a good event out of that in the transfer method since it only contains one side. In .find profile we already link ftReceiver using our wallet abstraction and I would love to be able to send in that capability to this to be used for transfer. 

Other use cases can be if you want to transfer funds from one wallet to a shared/pooled wallet of some kind. 

I would really love though to have a static function in the new standard that can emit a shared event. Right now when you want to crawl for deposit events you have to know all the contract names. If the shared interface contract has a static function (when we get those) that said you could use to transfer with then that function could emit an event and everything would be so much simple. (The same argument is even stronger in NFTs IMHO)

Reply:
I'm not understanding the point about the event. We could still have a shared event here whether or not we have a capability or address:
```
pub event TokensTransferred(amount: UFix64, from: Address?, to: Address?, type: Type)
```

I do see your point about wanting to be able to use different kinds of capabilities. So are you imagining a transaction like this?
```
// Can pass in any receiver path instead of just the default.
// This lets you choose the token you want to send as well the capability you want to send it to.
transaction(amount: UFix64, to: Address, senderPath: StoragePath, receiverPath: PublicPath) {

    // The Vault resource that holds the tokens that are being transferred
    let sentVault: @FungibleToken.Vault

    prepare(signer: AuthAccount) {

        // Get a reference to the signer's stored vault
        let vaultRef = signer.borrow<&{FungibleToken.Provider}>(from: senderPath)
			?? panic("Could not borrow reference to the owner's Vault!")

        let recipient = getAccount(to)
        let receiverCap = recipient.getCapability(receiverPath)

        // Transfer tokens from the signer's stored vault to the receiver capability
        vaultRef.transfer(amount: amount, recipient: receiverCap)
    }
}
```
That seems reasonable